optuna/storages/redis.py

```diff
@@ -0,0 +1,582 @@
+from optuna import exceptions
+from optuna.storages import base
+from optuna.storages.base import DEFAULT_STUDY_NAME_PREFIX
+from optuna import structs
+from optuna import type_checking
+
+try:
+    import redis
+
+    _available = True
+except ImportError as e:
+    _import_error = e
+    _available = False
+
+if type_checking.TYPE_CHECKING:
+    from typing import Any  # NOQA
+    from typing import Dict  # NOQA
+    from typing import List  # NOQA
+    from typing import Optional  # NOQA
+
+
+@experimental("1.3.0")
+class RedisStorage(base.BaseStorage):
+    """Storage class for Redis backend.
+
+    Note that library users can instantiate this class, but the attributes
+    provided by this class are not supposed to be directly accessed by them.
+
+    Example:
+
+        We create an :class:`~optuna.storages.redis.RedisStorage` instance using
+        the given redis database URL
+
+        .. code::
+
+            >>> import optuna
+            >>>
+            >>> def objective(trial):
+            >>>     ...
+            >>>
+            >>> storage = optuna.storages.redis.RedisStorage(
+            >>>     url='redis://passwd@localhost:port/db',
+            >>> )
+            >>>
+            >>> study = optuna.create_study(storage=storage)
+            >>> study.optimize(objective)
+
+    Args:
+        url: URL of the redis storage. password and db are optional
+                   (ie: redis://localhost:6379)
+
+    .. note::
+        If you use plan to use Redis as a storage mechanism for optuna,
+        make sure Redis in installed and running.
+        Please execute ``$ pip install -U redis`` to install redis python library.
+    """
+
+    def __init__(self, url):
+        # type: (str) -> None
+
+        if not _available:
+            raise ImportError(
+                "Redis in not available. Please install redis to use this feature. "
+                "Redis can be installed by executing `$ pip install -U redis`. "
+                "For further information, please refer to the installation guide of redis. "
+                "(The actual import error is as follows: " + str(_import_error) + ")"
+            )
+
+        self.redis = redis.Redis.from_url(url)
+
+    def create_new_study(self, study_name=None):
+        # type: (Optional[str]) -> int
+
+        key_study_name = "study_name:{}:study_id".format(study_name)
+        if study_name is not None and self.redis.exists(key_study_name):
+            raise exceptions.DuplicatedStudyError
+
+        if not self.redis.exists("study_counter"):
+            # we need the counter to start with 0
+            self.redis.set("study_counter", -1)
+        study_id = self.redis.incr("study_counter", 1)
+        # we need the trial_number counter to start with 0
+        self.redis.set("study_id:{:010d}:trial_number".format(study_id), -1)
+
+        if study_name is None:
+            study_name = "{}{:010d}".format(DEFAULT_STUDY_NAME_PREFIX, study_id)
```

Comment:
I added the following unit test to `tests/storages_tests/test_storages.py` on my local environment.
```python
@parametrize_storage
def test_create_and_delete_anonymous_study(storage_init_func):
    # type: (Callable[[], BaseStorage]) -> None

    storage = storage_init_func()
    study_id = storage.create_new_study()
    study_name = storage.get_study_name_from_id(study_id)
    assert study_id == storage.get_study_id_from_name(study_name)
```

This test was failed as follows:
```python
$ pytest tests/storages_tests/test_storages.py
...
=================================================================================================================== FAILURES ===================================================================================================================__________________________________________________________________________________________ test_create_and_delete_anonymous_study[redis_with_flushdb] __________________________________________________________________________________________

storage_init_func = <function redis_with_flushdb at 0x7fe733fbb620>

    @parametrize_storage
    def test_create_and_delete_anonymous_study(storage_init_func):
        # type: (Callable[[], BaseStorage]) -> None

        storage = storage_init_func()
        study_id = storage.create_new_study()
        study_name = storage.get_study_name_from_id(study_id)
>       assert study_id == storage.get_study_id_from_name(study_name)

tests/storages_tests/test_storages.py:112:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <optuna.storages.redis.RedisStorage object at 0x7fe711f7b748>, study_name = 'no-name-0000000000'

    def get_study_id_from_name(self, study_name):
        # type: (str) -> int

        if not self.redis.exists("study_name:{}:study_id".format(study_name)):
>           raise ValueError("No such study {}.".format(study_name))
E           ValueError: No such study no-name-0000000000.

optuna/storages/redis.py:218: ValueError
```

I think that the reassignment of `key_study_name` (i.e., `key_study_name = "study_name:{}:study_id".format(study_name)`) is needed after the updating of `study_name`.

Reply:
Good catch!. Sorted it out. Added the test as well.

Reply:
👍